xdg_shell.c

```diff
@@ -240,6 +294,19 @@ handle_xdg_toplevel_destroy(struct wl_listener *listener, void *data)
 	struct cg_xdg_shell_view *xdg_shell_view = wl_container_of(listener, xdg_shell_view, destroy);
 	struct cg_view *view = &xdg_shell_view->view;
 
+	if (xdg_shell_view->foreign_toplevel) {
+		wl_list_remove(&xdg_shell_view->foreign_toplevel->request_activate.link);
+		wl_list_remove(&xdg_shell_view->foreign_toplevel->request_close.link);
+		wl_list_remove(&xdg_shell_view->foreign_toplevel->request_fullscreen.link);
+		free(xdg_shell_view->foreign_toplevel);
+		xdg_shell_view->foreign_toplevel = NULL;
+	}
+
+	if (view->foreign_toplevel_handle) {
```

Comment:
This condition is unnecessary: `wlr_foreign_toplevel_handle_v1_destroy()` is a no-op if NULL is passed in (like `free()`).